vignettes/epiprocess.Rmd

```diff
@@ -198,3 +198,89 @@ ggplot(x, aes(x = time_value, y = cases)) +
   labs(x = "Date", y = "Confirmed cases of Ebola in Sierra Leone") 
 ```
+
+## Examples on Additional Keys in epi_df
+In the following examples we will show how to create an `epi_df` with additional keys.
+
+### Convert a `tsibble` that has county code as an extra key
+```{r}
+ex1 <- tibble(
+  geo_value = rep(c("ca", "fl", "pa"), each = 3),
+  county_code = c(06059,06061,06067,
+             12111,12113,12117,
+             42101, 42103,42105),
+  time_value = rep(seq(as.Date("2020-06-01"), as.Date("2020-06-03"),
+                       by = "day"), length.out = length(geo_value)),
+  value = 1:length(geo_value) + 0.01 * rnorm(length(geo_value))
+  ) %>% 
+  as_tsibble(index = time_value, key = c(geo_value, county_code))
+
+ex1 <- as_epi_df(x = ex1, geo_type = "state", time_type = "day", as_of = "2020-06-03")
+```
+
+The metadata now includes `county_code` as an extra key.
+```{r}
+attr(ex1,"metadata")
+```
+
+
+### Dealing with misspecified column names 
+
+`epi_df` requires there to be columns `geo_value` and `time_value`, if they do not exist then `as_epi_df()` throws an error.
+```{r, error = TRUE}
+data.frame(
+  state = rep(c("ca", "fl", "pa"), each = 3), # misnamed
+  pol = rep(c("blue", "swing", "swing"), each = 3), # extra key
+  reported_date = rep(seq(as.Date("2020-06-01"), as.Date("2020-06-03"),
+                       by = "day"), length.out = length(geo_value)), # misnamed
+  value = 1:length(geo_value) + 0.01 * rnorm(length(geo_value))
+) %>% as_epi_df() 
+```
+
+The columns can be renamed to match `epi_df` format. In the example below, notice there is also an additional key `pol`. 
+```{r}
+ex2 <- tibble(
+  state = rep(c("ca", "fl", "pa"), each = 3), # misnamed
+  pol = rep(c("blue", "swing", "swing"), each = 3), # extra key
+  reported_date = rep(seq(as.Date("2020-06-01"), as.Date("2020-06-03"),
+                          by = "day"), length.out = length(state)), # misnamed
+  value = 1:length(state) + 0.01 * rnorm(length(state))
+) %>% data.frame()
+
+head(ex2) 
+
+ex2 <- ex2 %>% rename(geo_value = state, time_value = reported_date) %>%
+  as_epi_df(geo_type = "state", as_of = "2020-06-03", 
+            additional_metadata = c(other_keys = "pol"))
+
+attr(ex2,"metadata")
+```
+
+
+### Adding additional keys to an `epi_df` object
+
+In the above examples, all the keys are added to objects that are not `epi_df` objects. We illustrate how to add keys to an `epi_df` object. 
+
+We use a subset dataset from the the `covidcast` library. 
+
+```{r}
+ex3 <- jhu_csse_county_level_subset %>%
+  filter(time_value > "2021-12-01", state_name == "Massachusetts") %>%
+  slice_tail(n = 6) 
+  
+attr(ex3,"metadata") # geo_type is county currently
+```
+
+Now we add state (MA) as a new column and a key to the metadata. 
+```{r}
+
+ex3 <- ex3 %>% 
+  as_tsibble() %>% # needed to add the additional metadata
+  mutate(state = rep("MA",6)) %>%
+  as_epi_df(additional_metadata = c(other_keys = "state"))
+
+attr(ex3,"metadata")
+```
+
+
+
```

Comment:
If it's not already done, explain how setting these other keys is important / changes anything?  (E.g., default `epi_slide` behavior.)

Reply:
Hi @brookslogan! 

I'm actually unsure about how the other keys will affect `epi_slide()`? Would you be able to elaborate a bit more on this?

For example, would we expect the default bahaviour of `epi_slide` to group by geo_value and the other keys? Currently, it doesn't seem to be doing that. 

``` r
suppressPackageStartupMessages(library(epiprocess))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(tsibble))

ex1 <- tibble(
  geo_value = rep(c("ca", "fl", "pa"), each = 3),
  county_code = c("06059","06061","06067",
                  "12111","12113","12117",
                  "42101","42103","42105"),
  time_value = rep(seq(as.Date("2020-06-01"), as.Date("2020-06-03"),
                       by = "day"), length.out = length(geo_value)),
  value = 1:length(geo_value) + 0.01 * withr::with_rng_version("3.0.0", withr::with_seed(42, length(geo_value)))
) %>% 
  as_tsibble(index = time_value, key = c(geo_value, county_code))

ex1 <- as_epi_df(x = ex1, geo_type = "state", time_type = "day", as_of = "2020-06-03")

# do we expect the following two lines to be the same?
ex1  %>% group_by(geo_value, county_code) %>% epi_slide(avg = mean(value), n = 2)
#> An `epi_df` object, with metadata:
#> * geo_type  = state
#> * time_type = day
#> * as_of     = 2020-06-03
#> 
#> # A tibble: 9 x 5
#> # Groups:   geo_value, county_code [9]
#>   geo_value county_code time_value value   avg
#> * <chr>     <chr>       <date>     <dbl> <dbl>
#> 1 ca        06059       2020-06-01  1.09  1.09
#> 2 ca        06061       2020-06-02  2.09  2.09
#> 3 ca        06067       2020-06-03  3.09  3.09
#> 4 fl        12111       2020-06-01  4.09  4.09
#> 5 fl        12113       2020-06-02  5.09  5.09
#> 6 fl        12117       2020-06-03  6.09  6.09
#> 7 pa        42101       2020-06-01  7.09  7.09
#> 8 pa        42103       2020-06-02  8.09  8.09
#> 9 pa        42105       2020-06-03  9.09  9.09
ex1  %>% epi_slide(avg = mean(value), n = 2)
#> An `epi_df` object, with metadata:
#> * geo_type  = state
#> * time_type = day
#> * as_of     = 2020-06-03
#> 
#> # A tibble: 9 x 5
#>   geo_value time_value county_code value   avg
#> * <chr>     <date>     <chr>       <dbl> <dbl>
#> 1 ca        2020-06-01 06059        1.09  4.09
#> 2 fl        2020-06-01 12111        4.09  4.09
#> 3 pa        2020-06-01 42101        7.09  4.09
#> 4 ca        2020-06-02 06061        2.09  4.59
#> 5 fl        2020-06-02 12113        5.09  4.59
#> 6 pa        2020-06-02 42103        8.09  4.59
#> 7 ca        2020-06-03 06067        3.09  5.59
#> 8 fl        2020-06-03 12117        6.09  5.59
#> 9 pa        2020-06-03 42105        9.09  5.59
```
<sup>Created on 2022-07-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
I wasn't sure either.  Looking over, I also don't see it impacting `epi_slide`.  And from a quick skim, it looks like the `"other_keys"` metadata in `epi_df` does not impact anything and is purely informational at this point, unlike in `as_epi_archive`, where the `other_keys` parameter affects how the update data is interpreted.  This mismatch may be surprising, so (a) we should document it, even if it's saying "we don't use this in one case", and (b) we may want to revisit it (see https://github.com/cmu-delphi/epiprocess/issues/89).

Reply:
Sure thing, thanks Logan. I'll add a one-line comment in the vignette to point out that currently `other_keys` metadata in `epi_df` doesn't impact `epi_slide`, contrary to `other_keys` in `as_epi_archive` which affects how the update data is interpreted.